core/unit_test/TestHalfOperators.hpp

```diff
@@ -887,7 +895,7 @@ struct Functor_TestHalfOperators {
     // expected_lhs(TW) = d_lhs <=> d_rhs;  // Need C++20?
 
     actual_lhs(PASS_BY_REF)   = static_cast<double>(accept_ref(h_lhs));
-    expected_lhs(PASS_BY_REF) = d_lhs;
+    expected_lhs(PASS_BY_REF) = accept_ref_expected(h_lhs);
```

Comment:
I am sorry I don't get that change.  Can you please confirm that the test fail without it.

Reply:
@dalg24: I have confirmed it fails without it. Please feel free to revert this change to confirm.

Reply:
ok I don't get it either. How are `d_lhs` and `accept_ref_expected(h_lhs)` different? Does this test pass if you just assign `h_lhs` instead of `accept_ref_exepcted(h_lhs)`?

Reply:
> Not blocking but also don't fully understand the changes in the unit test

@dalg24, @crtrott: Maybe my understanding is flawed. It appears that the compiler was optimizing out the casts float -> double -> float when we had:
```
expected_lhs(PASS_BY_REF) = d_lhs
```
 
From the compiler writer's perspective, this makes sense since there's no need to spend runtime cycles casting from float -> double -> float again during instantiation: https://github.com/kokkos/kokkos/pull/5778/files/15ed7c0bad93bf7d6040900eb498bfbcf71dcd3e#diff-6980b46765a569d9ae64dffdbe6a0db53b9ed49c8b9157236055a90add5b47afR268. Note that in this case half is an alias to float.

For the unit test though, this means that actual_lhs gets the result of the casts float -> double -> float which is accompanied by rounding while expected_lhs was not cast from float -> double -> float, instead, it was getting a store of float -> double. I believe this is the cause of the unit-test failures that we're previously erroneously addressed with the `epsilon_bias`. I believe this casting difference is propagated by the compiler "helping with optimizations" from the lines we're commenting on here as well as https://github.com/kokkos/kokkos/pull/5778/files/15ed7c0bad93bf7d6040900eb498bfbcf71dcd3e#diff-6980b46765a569d9ae64dffdbe6a0db53b9ed49c8b9157236055a90add5b47afR268. We didn't see this for half_t since half_t always maps to a half precision float on all the Kokkos backends that support half precision (HIP, CUDA, and SYCL). CUDA is the only backend that supports bhalf_t and only CUDA 11+ supports it. This is my understanding. Does that make sense? Do you agree?

Reply:
Also, I don't think this was failing when half_t or bhalf_t mapped to a 16-bit floating point type since the epsilons for those types are larger than FLT_EPSILON.